### ISSUE TITLE
sporadic sms error log messages - Cannot read property 'Average' of undefined

### ISSUE THREAD

**Issue by jrgm** [mozilla/fxa-auth-server#2428]
In stage, noticed these in the logs:

"Logger":"fxa-auth-server","Type":"sms.budget.error","Severity":2,"Pid":1,"EnvVersion":"2.0","Fields":{"op":"sms.budget.error","err":"Cannot read property 'Average' of undefined"}}

This is mainly FYI. The error is caught and the budget is assumed Ok. They don't happen all the time. I attribute this to low sms usage in stage. No action necessarily required (although a positive "I checked and it's ok because usage is X" log message would be useful for debugging purposes, if ever needed in the future.)

**Comment by philbooth** [comment]
So is the way this works that `cloudwatch.getMetricStatistics` only returns datapoints if there's been a change in the specified metric within the specified timeframe? If so, is it possible that the new disabling behaviour could get into an infinite failure loop?

The scenario I'm picturing:

1. Budget check detects that we're within $200 of the spend limit and disables SMS.
2. SMS is disabled for more than 5 minutes, which is the time period we specify to `getMetricStatistics`.
3. We get the budget increased.
4. Because SMS was disabled for more than 5 minutes, `getMetricStatistics` never returns a datapoint for us to check and so we never re-enable SMS.

Would we be better off specifying a longer time period than 5 minutes, or have I misunderstood the cause of the failed logging on stage?

**Comment by philbooth** [comment]
> ...is it possible that the new disabling behaviour could get into an infinite failure loop?

Whoops, that makes no sense given our logic. What would actually happen is that we'd immediately re-enable SMS as soon as we stop receiving datapoints from `getMetricStatistics`. Is *that* something we need to worry about?

**Comment by jrgm** [comment]
So, if we stopped receiving datapoints and assumed it was okay to continue, and then kept failing on sending, eventually that would trigger a page to ops to investigate. At which point, we could manually disable sms if, e.g., we were in fact over-budget.

So, I don't think we have to worry about getting snookered that badly on no data. But the positive log message would be useful.

**Comment by philbooth** [comment]
> So, I don't think we have to worry about getting snookered that badly on no data.

That's true but if the feature isn't going to do it's job properly I wonder if it's worth having in there at all. Would some basic retry logic make it more robust and worth keeping around?

So if the average over 5 minutes returns no datapoints and we previously disabled SMS, we retry the query with a period of an hour. Then if that fails we retry with a period of a day. Then if that fails we retry with a period of a week. Then if that fails we retry with whatever the period from the start of the month until now is. Then only if all those failed do we actually give up and re-enable SMS?

> But the positive log message would be useful.

Yep, I'll add this.
